parsl/data_provider/staging.py

```diff
@@ -0,0 +1,68 @@
+from concurrent.futures import Future
+from typing import Optional, Callable
+from parsl.app.futures import DataFuture
+from parsl.data_provider.files import File
+
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from parsl.data_provider.data_manager import DataManager
+
+
+class Staging:
+    """
+    This class defines the interface for file staging providers.
+
+    For each file to be staged in, the data manager will present the file
+    to each configured Staging provider in turn: first, it will ask if the
+    provider can stage this file by calling `can_stage_in`, and if so, it
+    will call both `stage_in` and `replace_task` to give the provider the
+    opportunity to perform staging.
+
+    For each file to be staged out, the data manager will follow the same
+    pattern using the corresponding stage out methods of this class.
+
+    The default implementation of this class rejects all files, and
+    performs no staging actions.
+
+    To implement a concrete provider, one or both of the `can_stage_*`
+    methods should be overridden to match the appropriate files, and then
+    the corresponding `stage_*` and/or `replace_task*` methods should be
+    implemented.
```

Comment:
I don't see `defaultStaging` documented anywhere user-facing. Maybe this is a good place to mention it?

Reply:
I think the right place is in an updated userguide data section which I'd like to write in a different PR

Reply:
I see an example has been added to the user guide. I still favor putting important bits in docstrings for users that like to stay in their terminal or notebook and do `help(foo)`. But I'm also ok with pressing forward as-is.